bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+        if (accessTokenResponse != null) {
+            // store it
+            storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+        }
+        return accessTokenResponse;
+
+    }
+
+    @Override
+    public AccessTokenResponse getAccessTokenResponse() throws OAuthException, IOException, OAuthError {
+        checkNotNull(handle, "handle must not be null");
+        AccessTokenResponse lastAccessToken = storeHandler.loadAccessTokenResponse(handle);
+        if (lastAccessToken == null) {
+            return null;
+        }
+
+        if (lastAccessToken.isExpired(new Date(), tokenExpiresIn_Buffer) && lastAccessToken.getRefreshToken() != null) {
+            return refreshToken();
+        }
+        return lastAccessToken;
+    }
+
+    @Override
+    public void importAccessTokenResponse(AccessTokenResponse accessTokenResponse) {
+        storeHandler.saveAccessTokenResponse(handle, accessTokenResponse);
+    }
+
+    /**
+     * Access tokens have expiry times. They are given by authorization server.
+     * This parameter introduces a buffer in seconds that deduct from the expires-in.
+     * For example, if the expires-in = 3600 seconds ( 1hour ), then setExpiresInBuffer(60)
+     * will remove 60 seconds from the expiry time. In other words, the expires-in
+     * becomes 3540 ( 59 mins ) effectively.
+     *
+     * Calls to protected resources can reasonably assume that the token is not expired.
+     *
+     * @param seconds The number of seconds to remove the expires-in. Default 0 seconds.
+     */
+    public void setTokenExpiresInBuffer(int tokenExpiresInBuffer) {
+        this.persistedParams.tokenExpiresInBuffer = tokenExpiresInBuffer;
+    }
+
+    @Override
+    public void remove() {
+        logger.debug("removing handle: {}", handle);
+        storeHandler.remove(handle);
+    }
+
+    @Override
+    public void close() {
+        closed = true;
+        storeHandler = null;
+        persistedParams = null;
+
+        logger.debug("closing oauth client, handle: {}", handle);
+        if (httpClientFactoryServiceRef != null) {
+            bundleContext.ungetService(httpClientFactoryServiceRef);
+        }
+    }
+
+    @Override
+    public boolean isClosed() {
+        return closed;
+    }
+
+    private void init() {
+        // Using OSGI R4.2 compatibility
+        httpClientFactoryServiceRef = bundleContext.getServiceReference(HttpClientFactory.class.getName());
+        httpClientFactory = (HttpClientFactory) bundleContext.getService(httpClientFactoryServiceRef);
+    }
+
+    private String createNewState() {
+        return UUID.randomUUID().toString();
+    }
+
+    // Params that need to be persisted
+    public static class PersistedParams {
+        String handle;
+        String tokenUrl;
+        String authorizationUrl;
+        String clientId;
+        String clientSecret;
+        String scope;
+        Boolean supportsBasicAuth;
```

Comment:
primitive boolean

Reply:
null value means don't know in here, rather than a direct yes/ no.